graylog2-server/src/main/java/org/graylog2/indexer/messages/ImmutableMessage.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.indexer.messages;
+
+import org.graylog2.indexer.IndexSet;
+import org.graylog2.plugin.Message;
+
+import java.util.Map;
+import java.util.Set;
+
+public interface ImmutableMessage extends Indexable, Acknowledgeable {
+
+    // Overriding Indexable#getId, because there it's deprecated, but in the context of a Message, it still makes sense
+    @Override
+    String getId();
+
+    static ImmutableMessage wrap(Message message) {
+        return new SerializationMemoizingMessage(message);
+    }
+
+    Set<IndexSet> getIndexSets();
+
+    Map<String, Object> getFields();
+
+    String getMessage();
+
+    Object getField(String key);
+
+    String getSource();
+
+    Set<String> getStreamIds();
```

Comment:
Should we return `ImmutableSet` and `ImmutableMap` here instead to make it clear that the returned maps and sets are immutable?

Reply:
Personally, I would always assume that maps or sets returned from a method call are immutable. Even more so with the abundance of `Map#of` and `Set#of` calls. I prefer avoiding the Guava types in an API but I'm undecided. Maybe a 3rd opinion about that, @kroepke?

Reply:
I usually also regard the interfaces as coming with no guarantees, but it would probably be clearer if we returned `Immutable*` types here as the entire class is somewhat special.
The guava collections implement `Map` and `Set`, so callers should have no issue using the results freely with other code.

Reply:
Allright, I'll do that then. Thanks for your input!